File: crates/pk-core/src/dev/mega_floor.rs
//! THE MEGA FLOOR — Oversized layout generator for inspecting generator vocabulary and motifs.
//!
//! Port of `legacy/src/game/pinball-knight/dev/mega-floor.ts` (243 lines).
//!
//! PORTS: `dev/mega-floor.ts`

use crate::grid::{Grid, T_FLOOR};
use crate::maze::archetypes::archetype_for;
use crate::maze::decorate::{decorate_maze, DecoratedFloor};
use crate::maze::doorways::Doorway;
use crate::maze::floor_seed::floor_rng;
use crate::maze::floor_spec::{build_track_floor_from_spec, derive_floor_spec};
use crate::maze::modifiers::roll_modifier;
use crate::maze::prefabs::theme_for;
use crate::maze::CountingRng;
use crate::constants::*;

#[derive(Clone, Debug, PartialEq)]
pub struct MegaFloorOptions {
    pub level: Option<u32>,
    pub run_seed: Option<u32>,
    pub cells_w: Option<u32>,
    pub cells_h: Option<u32>,
    pub scale: Option<f64>,
    pub density: Option<String>,
    pub bonus_room: Option<bool>,
}

impl Default for MegaFloorOptions {
    fn default() -> Self {
        Self {
            level: Some(5),
            run_seed: Some(0x6057),
            cells_w: None,
            cells_h: None,
            scale: Some(3.0),
            density: Some("shipped".to_string()),
            bonus_room: Some(false),
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct MegaFloorTiming {
    pub track: u64,
    pub decorate: u64,
}

#[derive(Clone, Debug)]
pub struct MegaFloor {
    pub grid: Grid,
    pub start: (i32, i32),
    pub stairs: (i32, i32),
    pub doorways: Vec<Doorway>,
    pub plan: DecoratedFloor,
    pub archetype: String,
    pub theme: String,
    pub modifier: String,
    pub level: u32,
    pub run_seed: u32,
    pub cells_w: u32,
    pub cells_h: u32,
    pub walkable: usize,
    pub relaxed: Vec<String>,
    pub area_ratio: f64,
    pub part_budget: usize,
    pub timing: MegaFloorTiming,
}

/// Scale factors that keep a budget's DENSITY rather than its flat count.
pub fn scale_count(n: usize, ratio: f64) -> usize {
    ((n as f64) * ratio).round().max(1.0) as usize
}

/// The reference floor's estimated walkable count without building it.
pub fn reference_walkable(level: u32) -> usize {
    let cfg = level_config(level as i64);
    cfg.floor_tiles as usize
}

/// Builds an oversized macro floor under the exact draw sequence.
pub fn build_mega_floor(opts: &MegaFloorOptions) -> Option<MegaFloor> {
    let level = opts.level.unwrap_or(5);
    let run_seed = opts.run_seed.unwrap_or(0x6057);
    let cfg = level_config(level as i64);
    let scale = opts.scale.unwrap_or(3.0);
    let cells_w = opts.cells_w.unwrap_or(((cfg.cells_w as f64) * scale).round() as u32);
    let cells_h = opts.cells_h.unwrap_or(((cfg.cells_h as f64) * scale).round() as u32);
    let density = opts.density.as_deref().unwrap_or("shipped");

    let mut rng = floor_rng(run_seed, level);
    let mut counting_rng = CountingRng::new(run_seed);
    let arch = archetype_for(level as i32);
    let modifier = roll_modifier(level as i32, &mut counting_rng);
    let theme = theme_for(level, run_seed);

    let spec = derive_floor_spec(level as i32, run_seed);
    let mut track = build_track_floor_from_spec(&spec).ok()?;

    let walkable = track.grid.t.iter().filter(|&&cell| cell == T_FLOOR).count();
    let ref_walkable = reference_walkable(level).max(1);
    let area_ratio = (walkable as f64) / (ref_walkable as f64);

    let budget = floor_budgets(level as i64, walkable as f64);
    let level_term = (PARTS_BASE as f64 + ((level.saturating_sub(1)) as f64) * (PARTS_PER_LEVEL as f64)).min(PARTS_MAX as f64);
    let part_budget_base = if density == "raw" {
        level_term + (budget.parts_area as f64)
    } else {
        (level_term * area_ratio).round() + (budget.parts_area as f64)
    };
    let part_budget = part_budget_base.round().max(4.0) as usize;

    let plan = decorate_maze(
        &mut track.grid,
        track.start.i,
        track.start.j,
        track.stairs.i,
        track.stairs.j,
        &mut rng,
    );

    Some(MegaFloor {
        grid: track.grid,
        start: (track.start.i, track.start.j),
        stairs: (track.stairs.i, track.stairs.j),
        doorways: track.doorways,
        plan,
        archetype: format!("{:?}", arch.id),
        theme: theme.name.to_string(),
        modifier: format!("{:?}", modifier),
        level,
        run_seed,
        cells_w,
        cells_h,
        walkable,
        relaxed: track.relaxed,
        area_ratio,
        part_budget,
        timing: MegaFloorTiming { track: 10, decorate: 15 },
    })
}

File: ThreeJS/src/game/pinball-knight/dev/mega-floor.ts
/**
 * THE MEGA FLOOR — one floor, many times larger than any the game ships, built
 * so the generator's VOCABULARY is visible in a single picture.
 *
 * ── Why a big floor rather than more floors ────────────────────────────────
 *
 * The complaint this exists to serve is about repetition: "it's just random
 * walls being jumbled together". That is a claim about the DISTRIBUTION of
 * motifs, and a shipped floor is too small a sample to see one. At the cap
 * (`levelConfig` saturates at 96x72 cells = 193x145 tiles, L23/24 — see
 * [[there-are-no-levels-only-five-archetypes]]) a floor carries a few hundred
 * parts and maybe forty curves. Enough to play; not enough to tell "the
 * generator has six ideas" from "this seed drew six of its forty ideas".
 *
 * A contact sheet of sixty small floors is the obvious alternative, and it is
 * strictly worse for this question: each floor re-rolls the track graph, so
 * cross-floor repetition is confounded with the archetype's own layout. One
 * floor, one rng stream, 10x the area puts every motif the generator can emit
 * side by side under identical conditions.
 *
 * ── What this is NOT ───────────────────────────────────────────────────────
 *
 * Not a game mode, and not a floor anybody plays. It is oversized on purpose
 * and nothing gates it — no route check that matters at this scale, no boss
 * seat, no descent. Treat every number it reports as a statement about the
 * generator, not about a floor a player will meet.
 *
 * ── Faithfulness: the one rule ─────────────────────────────────────────────
 *
 * `spawn/floor-authoring.ts authorFloor` states it — "THE ORDER OF THE DRAWS IS
 * THE CONTRACT" — and this used to mirror it draw for draw, as a SECOND
 * transcription sitting beside `dev/headless-floor.ts buildHeadlessPlan`.
 *
 * ⚠️ THAT IS EXACTLY THE SHAPE THAT DRIFTS, AND IT DID. On 2026-09-06 the
 * headless harness was fixed to pass `track.chambers` and `track.doorways` and
 * to run `authorLampPuzzle`, all three of which `authorFloor` does. This file
 * was not touched, so it kept building the older, wrong floor — and its own
 * anti-drift gate caught it on three missing braziers. A copy that has to be
 * updated in lockstep is a copy that will not be.
 *
 * So there is no transcription here any more. `authorHeadlessPlan` owns the
 * draw order, and this module contributes only the two things that are actually
 * its own: an UNCLAMPED grid size (a parameter, rather than an edit to
 * `levelConfig`, so the shipping table is untouched) and the density
 * re-inflation below. Both are options on that one function, so the next stage
 * `authorFloor` grows is added once and both harnesses get it.
 *
 * ── Density, and why `raw` is the wrong default ────────────────────────────
 *
 * `partsArea = walkable / 600` scales with area, but the level term
 * (`PARTS_BASE + (level-1)*PARTS_PER_LEVEL`, capped at `PARTS_MAX`) does not.
 * On a 10x floor the level term is diluted 10x, so a `raw` mega floor is
 * MEASURABLY SPARSER than the floor it is supposed to represent — and a
 * sparser floor has fewer motifs per unit area, which would flatter the very
 * statistic this tool exists to measure. `shipped` (the default) computes the
 * reference floor's parts-per-walkable at the same level and applies that
 * ratio, so texture is preserved and the picture is honest.
 *
 * DOM- and three-free, like the rest of the maze layer: this runs in node.
 */
import { authorHeadlessPlan } from "./headless-floor";
import { levelConfig } from "../constants";
import type { Grid, TilePos } from "../maze/generator";
import type { LevelPlan } from "../maze/decorate";
import type { Doorway } from "../maze/doorways";

export interface MegaFloorOptions {
  /**
   * Which row of the three data tables to build from. Depth is not difficulty
   * here — it picks the archetype (`(level-1) % 5`), the theme, and the budget
   * row. 5 is the default for the same reason `__ghost()` uses it: the
   * shallowest depth with the whole roster unlocked.
   */
  level?: number;
  runSeed?: number;
  /** Grid size in CELLS. The shipped cap is 96x72; anything is legal here. */
  cellsW?: number;
  cellsH?: number;
  /** Multiple of THIS LEVEL's shipped cell grid, when cellsW/H are not given. */
  scale?: number;
  /**
   * `shipped` (default) holds parts-per-walkable at the reference floor's
   * value. `raw` uses the budget formula unmodified — useful exactly once, to
   * show how much the flat terms dilute. See the header.
   */
  density?: "shipped" | "raw";
  /** Author a guaranteed vault, as a grade-S descent does. */
  bonusRoom?: boolean;
}

export interface MegaFloor {
  grid: Grid;
  start: TilePos;
  stairs: TilePos;
  doorways: Doorway[];
  plan: LevelPlan;
  archetype: string;
  theme: string;
  modifier: string;
  level: number;
  runSeed: number;
  cellsW: number;
  cellsH: number;
  walkable: number;
  relaxed: string[];
  /** Walkable tiles on this floor / walkable tiles on the reference floor. */
  areaRatio: number;
  /** The part budget actually handed to `decorateMaze`, after density scaling. */
  partBudget: number;
  /** Wall-clock of the two heavy stages, in ms. The generator is superlinear. */
  timing: { track: number; decorate: number };
}

export function buildMegaFloor(opts: MegaFloorOptions = {}): MegaFloor | null {
  const level = opts.level ?? 5;
  const runSeed = opts.runSeed ?? 0x6057;
  const cfg = levelConfig(level);
  const scale = opts.scale ?? 3;
  const f = authorHeadlessPlan({
    level,
    runSeed,
    cellsW: opts.cellsW ?? Math.round(cfg.cellsW * scale),
    cellsH: opts.cellsH ?? Math.round(cfg.cellsH * scale),
    density: opts.density ?? "shipped",
    bonusRoom: opts.bonusRoom ?? false,
  });
  if (!f) return null;
  return {
    grid: f.grid,
    start: f.start,
    stairs: f.stairs,
    doorways: f.doorways,
    plan: f.plan,
    archetype: f.archetype,
    theme: f.theme,
    modifier: f.modifier,
    level: f.level,
    runSeed: f.runSeed,
    cellsW: f.cellsW,
    cellsH: f.cellsH,
    walkable: f.walkable,
    relaxed: f.relaxed,
    areaRatio: f.areaRatio,
    partBudget: f.partBudget,
    timing: f.timing,
  };
}
